rfc/system/0000-simplify-retry-behaviour.md

```diff
@@ -0,0 +1,120 @@
+* If preemptions counted against the user retry budget, we would have to switch to non-preemptible instances for the last retry of the user retry budget [instead of the system retry budget](https://github.com/flyteorg/flytepropeller/blob/a3c6e91f19c19601a957b29891437112868845de/pkg/controller/nodes/node_exec_context.go#L213).
+
+The downside of this approach is that the general retry behaviour remains as complex and difficult to understand as it is now.
+
+### 6.2 Demistify failure of Pods belonging to plugins/CRDs
+
+We discussed ways to give plugins access to the status of not only the custom resource they create (e.g. `PyTorchJob`) but also to the statuses of the pods created from the custom resource by the respective third-party operator (e.g. the kubeflow training operator). (This might require changes to the plugin interface, see details [here](https://github.com/flyteorg/flyte/discussions/3793).)
+
+This approach would allow plugins to demistify failures of tasks by looking at the underlying pods which, in contrast to the custom resource, do contain hints of preemptions. This way, we could unify the retry behaviour upon preemptions of the different plugins. However, wouldn't make the retry behaviour easier to understand and users still wouldn't have control over the number of preemptions which is why we opted to not continue these discussions either.
+
+
+### 6.3 Inform the user which retry budget an attempt counted against
+
+We considered showing the user in the Flyte console whether an attempt counted against the user retry budget or the system retry budget.
+
+This only improves the transparency and does not remove the different behavior between plugins. Doing so is not necessary when every failure is counted against the same retry budget.
+
+![budgets](https://github.com/flyteorg/flyte/assets/26092524/3e425528-ce3f-42af-8cc3-f0ca0cdab660)
+
+
+
+## 7 Potential Impact and Dependencies
+
+By having to activate the new retry behaviour in the platform configuration, we will not break existing workflows.
+
+## 8 Unresolved questions
+
+An open question is how we deal with the `interruptibleFailureThreshold`. Currently, this value is a positive integer that is the number of retries that are interruptible. For example, 4 (system failure) retries are allowed where 3 of them are interruptible. Defining it this way does not make as much sense when we allow for task-level configuration of the retries.
+
+We could define the number of retries that ARE NOT interruptible. For example a value of `nonInterruptibleRetries` that is set to 1 means that the last retry is not interruptible regardless of how many retries there are. We also have to discuss whether this value would be configured on the platform side (only for the new retry behaviour) or whether we would add this value to to task-level metadata and expose it to users. Adding this flag while ensuring backwards compatibility will require further discussion.
+
```

Comment:
Rather something we need to ensure/test: after the change non-retryable exceptions in the user code should not trigger a retry - even when using the PytorchJob plugin.

Probably not worth mentioning? Or do you think this might break?

Reply:
https://github.com/flyteorg/flyte/pull/3902/commits/8a9409aee245881aa83c7769e5afede8229ff55f

Good point! I don't think it hurts to be explicit here.